### PR TITLE
Remove conditional dependencies

### PR DESCRIPTION
**Pull request by tcollier**
@kristjan @jrichard @heidi 

Force the client to explicitly require the back-end library (i.e. ActiveJob or Dynosaur). Before this gem was including both, so a client was pulling in more than it needed.

The bulk of the change log is made up of moving files around (e.g. all Dynosaur-specific code is now namespaced in `BulkProcessor::BackEnd::Dynosaur`) and pushing all argument serialization into the back-end implementations, rather than in the generic `BackEnd` helper module.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,6 +50,14 @@ including `:resque`.
 
 #### Back end: Dynosaur
 
+Include the `dynosau` gem in your Gemfile, e.g.
```
Comment (kristjan) [inline README.md]: :book: dynosaur

File: lib/bulk_processor/back_end/active_job.rb [hunk 2]
```diff
@@ -1,20 +1,25 @@
...
       def start
-        Job::ProcessCSV.perform_later(processor_class.name, payload, key)
+        ActiveJob::ProcessCSVJob.perform_later(processor_class, payload, key)
       end
 
       def split(num_processes)
-        Job::SplitCSV.perform_later(processor_class.name, payload,
-                                    key, num_processes)
+        ActiveJob::SplitCSVJob.perform_later(processor_class, payload, key,
+                                             num_processes.to_s)
```
Comment (kristjan) [inline lib/bulk_processor/back_end/active_job.rb]: Why the `to_s`? An integer is the right type here, and it seems like `SplitCSVJob.perform_later` should be responsible for making it a string if it needs to.

File: lib/bulk_processor/back_end/dynosaur/tasks.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+require_relative 'process_csv_task'
+require_relative 'split_csv_task'
```
Comment (kristjan) [inline lib/bulk_processor/back_end/dynosaur/tasks.rb]: The `dynosaur` version gets this init file, but the `active_job` version loaded its pair directly. We should be consistent for sake of readers.

File: spec/bulk_processor/back_end/dynosaur/split_csv_task_spec.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+require 'bulk_processor/back_end/dynosaur/split_csv_task'
+
+describe 'bulk_processor namespace' do
+  describe 'bulk_processor:split' do
+    let(:performer) { instance_double(BulkProcessor::SplitCSV, perform: true) }
+    let(:rake_cmd) do
+      'bulk_processor:split[MockCSVProcessor,foo=bar,file.csv,2]'
+    end
+
+    before(:all) { Rake::Task.define_task(:environment) }
```
Comment (kristjan) [inline spec/bulk_processor/back_end/dynosaur/split_csv_task_spec.rb]: So for someone to use this gem, they're going to need a `rake environment` task. That might be a nice requirement to remove in the future.

**Comment by kristjan** [comment]
LGTM

File: spec/bulk_processor/back_end/dynosaur/split_csv_task_spec.rb (hunk 4 above)
Comment (tcollier) [inline spec/bulk_processor/back_end/dynosaur/split_csv_task_spec.rb]: Yes. This is the same solution we use in `Redshifter`. I will document this requirement.

File: lib/bulk_processor/back_end/dynosaur/tasks.rb (hunk 3 above)
Comment (tcollier) [inline lib/bulk_processor/back_end/dynosaur/tasks.rb]: The difference is that a client using the `dynosaur` back-end needs to explicitly require the rake tasks in `Rakefile`. Having this file allows us to update the tasks as need be without having the client change the require line in the app.

**Pull request merged**